README.rst
audio-display
=============

**audio-display** is a set of utility aimed at rendering images based on audio input.
It aims at generating images to produce visual "companion" to audio files. Typically to create
a video clip supporting a musical composition.


fft2png
-------

**fft2png** is a command line utility to create a set of image files representing audio spectrum from a wav file.
It is an offline version of a spectrum analyser output and is not totally dissimilar to `Sonic Candle`_.

Generated files can be imported and postprocessed in any video edition tool accepting a set of images as input.
Just make sure that the framerate used during the call to **fft2png** matches the framerate at which you consume
images in your video edition software. Libre software able to consume and use those images includes, but aren't
limited to, Natron_.

You can also use FFmpeg_ for either previewing quickly or for simple needs (just the spectrum over a fixed image
or a video. Some samples of `ffmpeg usage`_ can be found later.


Usage
.....

    usage: fft2png [-h] [-d] [-n] [-r TARGET_FPS] [-R {0,1,2,3}] [-v]
                   [-w BAR_WIDTH] [-s BAR_SPACING] [-c BAR_COUNT] [-C COLOR]
                   [-b BLENDING] [-W FFT_WINDOW] [--image-height IMAGE_HEIGHT]
                   [--audio-min-freq AUDIO_MIN_FREQ]
                   [--audio-max-freq AUDIO_MAX_FREQ]
                   [--silence-ceiling SILENCE_CEILING] [-i INPUT_FILENAME] -o
                   OUTPUT_FILENAME_MASK

    GPL v3+ 2015 Olivier Jolly

    optional arguments:
      -h, --help            show this help message and exit
      -d, --debug           debug operations [default: False]
      -n                    don't perform actions [default: False]
      -r TARGET_FPS, --framerate TARGET_FPS
                            output framerate [default: 30]

      -R {0,1,2,3}, --renderer {0,1,2,3}
                            which renderer to use to display bars (0=filled, 1=hollow, 2=symetrical filled, 3=symetrical hollow)

      -v, --version         show program's version number and exit
      -w BAR_WIDTH, --bar-width BAR_WIDTH
                            bar width in output images
      -s BAR_SPACING, --bar-spacing BAR_SPACING
                            bar spacing in output images
      -c BAR_COUNT, --bar-count BAR_COUNT
                            number of bars in output images
      -C COLOR, --color COLOR
                            hexa color of bars [RRGGBB or RRGGBBAA, default: FFFFFFFF]
      -b BLENDING, --blending BLENDING
                            blending of previous spectrum into current one (0 =
                            display only fresh data, 1 = use as many previous than
                            fresh data)
      -W FFT_WINDOW, --window FFT_WINDOW
                            window size for FFT [default: 4096]
      --image-height IMAGE_HEIGHT
                            output images height
      --audio-min-freq AUDIO_MIN_FREQ
                            min frequency in input audio
      --audio-max-freq AUDIO_MAX_FREQ
                            max frequency in input audio
      --silence-ceiling SILENCE_CEILING
                            opposite of threshold considered silence [in dB,
                            default: 70]
      -i INPUT_FILENAME     input file in wav format
      -o OUTPUT_FILENAME_MASK
                            output filename mask (should contain {:06} or similar
                            to generate sequence)

    Convert audio file to stack of images

**-r**
  Framerate of the generated images. Should match the framerate at which they will be consumed.
  Higher framerate gives a smoother result.

**-R**
  Aspect of the bar representing power for one frequency. 0 uses filled boxes, 1 uses hollow boxes,
  2 uses filled boxes vertically centered and 3 uses hollow boxes vertically centered.

**-w**
  Width (in pixel) per bar.

**-s**
  Spacing (in pixel) between bars.

**-c**
  Number of bars per images.

**-C**
  Color of the bars in hexa. Can be RGB or RGBA. For instance, FF0000 will render pure opaque red bars,
  00FF0080 will render 50% transparent pure green bars, ...

**-b**
  Blending ratio from previous frame into current one. When set to 0, only fresh data will be used to
  render bars. When set to 1, bars will be rendered from an average of the fresh and previous frame data.
  Intermediate values will inject a fraction of the previous frame data into the current one for rendering.
  Lower values tends to render more reactive spectrum while higher ones will smooth data over time and react slower.

**-W**
  Spectrum generation window is the amount of data in the audio file used to determine the spectrum raw data.
  Lower value will make spectrum blockier but will be slightly faster to generate.

****

FFMpeg usage
............

.. _ffmpeg usage:

Installation
------------

**audio-display** will be installable from PyPI with a single pip command::

    pip install audio-display

Alternatively, **audio-display** can be run directly from sources after a git pull (recommended if you want to tweak
or read the source)::

    git clone https://gitlab.com/zeograd/audio-display.git
    cd audio-display && python setup.py install

or directly from its git repository::

    pip install git+https://gitlab.com/zeograd/audio-display.git

.. _Sonic Candle: http://soniccandle.sourceforge.net/
.. _Natron: http://natron.fr
.. _FFmpeg: http://ffmpeg.org


CHANGELOG
### Added
- initial fft2png version
- rendering using 4 drawing methods

### Changed


audio_display/wavfile.py
"""
Module to read / write wav files using numpy arrays

Functions
---------
`read`: Return the sample rate (in samples/sec) and data from a WAV file.

`write`: Write a numpy array as a WAV file.

"""
from __future__ import division, print_function, absolute_import

import sys
import numpy
import struct
import warnings


class WavFileWarning(UserWarning):
    pass

_big_endian = False

WAVE_FORMAT_PCM = 0x0001
WAVE_FORMAT_IEEE_FLOAT = 0x0003
WAVE_FORMAT_EXTENSIBLE = 0xfffe
KNOWN_WAVE_FORMATS = (WAVE_FORMAT_PCM, WAVE_FORMAT_IEEE_FLOAT)

# assumes file pointer is immediately
#  after the 'fmt ' id


def _read_fmt_chunk(fid):
    if _big_endian:
        fmt = '>'
    else:
        fmt = '<'
    res = struct.unpack(fmt+'iHHIIHH', fid.read(20))
    size, comp, noc, rate, sbytes, ba, bits = res
    if comp not in KNOWN_WAVE_FORMATS or size > 16:
        comp = WAVE_FORMAT_PCM
        warnings.warn("Unknown wave file format", WavFileWarning)
        if size > 16:
            fid.read(size - 16)

    return size, comp, noc, rate, sbytes, ba, bits


# assumes file pointer is immediately
#   after the 'data' id
def _read_data_chunk(fid, comp, noc, bits, mmap=False):
    if _big_endian:
        fmt = '>i'
    else:
        fmt = '<i'
    size = struct.unpack(fmt, fid.read(4))[0]

    bytes = bits//8
    if bits == 8:
        dtype = 'u1'
    else:
        if _big_endian:
            dtype = '>'
        else:
            dtype = '<'
        if comp == 1:
            dtype += 'i%d' % bytes
        else:
            dtype += 'f%d' % bytes
    if not mmap:
        data = numpy.fromstring(fid.read(size), dtype=dtype)
    else:
        start = fid.tell()
        data = numpy.memmap(fid, dtype=dtype, mode='c', offset=start,
                            shape=(size//bytes,))
        fid.seek(start + size)

    if noc > 1:
        data = data.reshape(-1, noc)
    return data


def _skip_unknown_chunk(fid):
    if _big_endian:
        fmt = '>i'
    else:
        fmt = '<i'

    data = fid.read(4)
    # call unpack() and seek() only if we have really read data from file
    # otherwise empty read at the end of the file would trigger
    # unnecessary exception at unpack() call
    # in case data equals somehow to 0, there is no need for seek() anyway
    if data:
        size = struct.unpack(fmt, data)[0]
        fid.seek(size, 1)


def _read_riff_chunk(fid):
    global _big_endian
    str1 = fid.read(4)
    if str1 == b'RIFX':
        _big_endian = True
    elif str1 != b'RIFF':
        raise ValueError("Not a WAV file.")
    if _big_endian:
        fmt = '>I'
    else:
        fmt = '<I'
    fsize = struct.unpack(fmt, fid.read(4))[0] + 8
    str2 = fid.read(4)
    if (str2 != b'WAVE'):
        raise ValueError("Not a WAV file.")
    if str1 == b'RIFX':
        _big_endian = True
    return fsize

# open a wave-file


def read(filename, mmap=False):
    """
    Return the sample rate (in samples/sec) and data from a WAV file

    Parameters
    ----------
    filename : string or open file handle
        Input wav file.
    mmap : bool, optional
        Whether to read data as memory mapped.
        Only to be used on real files (Default: False)

        .. versionadded:: 0.12.0

    Returns
    -------
    rate : int
        Sample rate of wav file
    data : numpy array
        Data read from wav file

    Notes
    -----
    * The file can be an open file or a filename.
    * The returned sample rate is a Python integer.
    * The data is returned as a numpy array with a data-type determined
      from the file.
    * This function cannot read wav files with 24 bit data.

    """
    if hasattr(filename, 'read'):
        fid = filename
        mmap = False
    else:
        fid = open(filename, 'rb')

    try:
        fsize = _read_riff_chunk(fid)
        noc = 1
        bits = 8
        comp = WAVE_FORMAT_PCM
        while (fid.tell() < fsize):
            # read the next chunk
            chunk_id = fid.read(4)
            if chunk_id == b'fmt ':
                size, comp, noc, rate, sbytes, ba, bits = _read_fmt_chunk(fid)
                if bits == 24:
                    raise ValueError("Unsupported bit depth: the wav file "
                                     "has 24 bit data.")
            elif chunk_id == b'fact':
                _skip_unknown_chunk(fid)
            elif chunk_id == b'data':
                data = _read_data_chunk(fid, comp, noc, bits, mmap=mmap)
            elif chunk_id == b'LIST':
                # Someday this could be handled properly but for now skip it
                _skip_unknown_chunk(fid)
            else:
                warnings.warn("Chunk (non-data) not understood, skipping it.",
                              WavFileWarning)
                _skip_unknown_chunk(fid)
    finally:
        if not hasattr(filename, 'read'):
            fid.close()
        else:
            fid.seek(0)

    return rate, data

# Write a wave-file
# sample rate, data


def write(filename, rate, data):
    """
    Write a numpy array as a WAV file

    Parameters
    ----------
    filename : string or open file handle
        Output wav file
    rate : int
        The sample rate (in samples/sec).
    data : ndarray
        A 1-D or 2-D numpy array of either integer or float data-type.

    Notes
    -----
    * The file can be an open file or a filename.

    * Writes a simple uncompressed WAV file.
    * The bits-per-sample will be determined by the data-type.
    * To write multiple-channels, use a 2-D array of shape
      (Nsamples, Nchannels).

    """
    if hasattr(filename, 'write'):
        fid = filename
    else:
        fid = open(filename, 'wb')

    try:
        dkind = data.dtype.kind
        if not (dkind == 'i' or dkind == 'f' or (dkind == 'u' and
                                                 data.dtype.itemsize == 1)):
            raise ValueError("Unsupported data type '%s'" % data.dtype)

        fid.write(b'RIFF')
        fid.write(b'\x00\x00\x00\x00')
        fid.write(b'WAVE')
        # fmt chunk
        fid.write(b'fmt ')
        if dkind == 'f':
            comp = 3
        else:
            comp = 1
        if data.ndim == 1:
            noc = 1
        else:
            noc = data.shape[1]
        bits = data.dtype.itemsize * 8
        sbytes = rate*(bits // 8)*noc
        ba = noc * (bits // 8)
        fid.write(struct.pack('<ihHIIHH', 16, comp, noc, rate, sbytes,
                              ba, bits))
        # data chunk
        fid.write(b'data')
        fid.write(struct.pack('<i', data.nbytes))
        if data.dtype.byteorder == '>' or (data.dtype.byteorder == '=' and
                                           sys.byteorder == 'big'):
            data = data.byteswap()
        _array_tofile(fid, data)

        # Determine file size and place it in correct
        #  position at start of the file.
        size = fid.tell()
        fid.seek(4)
        fid.write(struct.pack('<i', size-8))

    finally:
        if not hasattr(filename, 'write'):
            fid.close()
        else:
            fid.seek(0)


if sys.version_info[0] >= 3:
    def _array_tofile(fid, data):
        # ravel gives a c-contiguous buffer
        fid.write(data.ravel().view('b').data)
else:
    def _array_tofile(fid, data):
        fid.write(data.tostring())


audio_display/fft2png.py
# fft2png, Generate spectrum from audio to png
# Copyright (C) 2016  Olivier Jolly
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import argparse
import logging
import math
import sys

import numpy as np
import os
from . import wavfile
from PIL import Image
from PIL import ImageDraw

__all__ = []
__version__ = 0.5
__date__ = '2015-12-18'
__updated__ = '2016-03-17'
__author__ = 'olivier@pcedev.com'

class SpectrumWriter(object):
    def __init__(self, opts):
        self.bucket_nb = opts.bar_count
        self.bucket_pixel_spacing = opts.bar_spacing
        self.bucket_pixel_width = opts.bar_width
        self.height = opts.image_height

        self.color = opts.color

        self.min_freq = opts.audio_min_freq
        self.max_freq = opts.audio_max_freq

        self.silence_ceiling = opts.silence_ceiling

        self.output_filename_mask = opts.output_filename_mask

    def write_bar(self, bucket_start, line_data):
        raise NotImplementedError

    def write_spectrum(self, frequencies, spectrum, frame_index):

        self.image = Image.new("RGBA",
                               (self.bucket_nb * (self.bucket_pixel_spacing + self.bucket_pixel_width), self.height))
        self.draw = ImageDraw.Draw(self.image)

        spectrum_len = len(spectrum)

        # on last frame, it might happen that len is 0, just exit in this case
        if spectrum_len == 0:
            return

        display_freq = np.logspace(np.log(self.min_freq) / np.log(3),
                                   np.log(self.max_freq) / np.log(3),
                                   self.bucket_nb, base=3)

        interpolated_spectrum = np.interp(display_freq, frequencies, spectrum, left=0, right=0)

        for bucket_idx in range(self.bucket_nb):

            try:
                attenuation = 10 * np.log(interpolated_spectrum[bucket_idx] / spectrum_len)
                line_data = max(self.height + (self.height * attenuation / self.silence_ceiling), 1)
                logging.debug("freq %f, %f dB", display_freq[bucket_idx], attenuation)
            except ValueError:
                # on last frame, we might not have enough data to compute the average in a bucket
                continue

            if line_data:
                bucket_start = bucket_idx * (self.bucket_pixel_spacing + self.bucket_pixel_width)
                self.write_bar(bucket_start, line_data)
                # draw.rectangle((bucket_start, height, bucket_start + bucket_pixel_width, - line_data), fill=(255, 255, 255))

        self.image.save(self.output_filename_mask.format(frame_index))

        del self.draw


class FilledRectangleSpectrumWriter(SpectrumWriter):
    def write_bar(self, bucket_start, line_data):
        self.draw.rectangle(
            (bucket_start, self.height, bucket_start + self.bucket_pixel_width, self.height - line_data),
            fill=self.color)


class HollowRectangleSpectrumWriter(SpectrumWriter):
    def write_bar(self, bucket_start, line_data):
        self.draw.rectangle(
            (bucket_start, self.height, bucket_start + self.bucket_pixel_width, self.height - line_data),
            outline=self.color)


class SymetricalFilledRectangleSpectrumWriter(SpectrumWriter):
    def write_bar(self, bucket_start, line_data):
        self.draw.rectangle((bucket_start, self.height / 2 + line_data / 2, bucket_start + self.bucket_pixel_width,
                             self.height / 2 - line_data / 2), fill=self.color)


class SymetricalHollowRectangleSpectrumWriter(SpectrumWriter):
    def write_bar(self, bucket_start, line_data):
        self.draw.rectangle((bucket_start, self.height / 2 + line_data / 2, bucket_start + self.bucket_pixel_width,
                             self.height / 2 - line_data / 2), outline=self.color)


RENDERERS = [FilledRectangleSpectrumWriter,
             HollowRectangleSpectrumWriter,
             SymetricalFilledRectangleSpectrumWriter,
             SymetricalHollowRectangleSpectrumWriter,
             ]

def smooth_spectrum(spectrum, previous_spectrum, alpha=0):
    try:
        return (spectrum + alpha * previous_spectrum) / (1 + alpha)
    except (ValueError, TypeError):
        return spectrum


def compute_frequencies(spectrum, fs):
    return np.arange(spectrum.size) * (fs / 2) / (spectrum.size - 1)


def main(argv=None):
    program_name = os.path.basename(sys.argv[0])
    program_version = "v" + str(__version__)
    program_build_date = "%s" % __updated__

    program_version_string = '%%prog %s (%s)' % (program_version, program_build_date)
    program_longdesc = '''Convert audio file to stack of images'''
    program_license = "GPL v3+ 2015 Olivier Jolly"

    if argv is None:
        argv = sys.argv[1:]

    try:
        parser = argparse.ArgumentParser(epilog=program_longdesc,
                                         description=program_license)
        parser.add_argument("-d", "--debug", dest="debug", action="store_true",
                            default=False,
                            help="debug operations [default: %(default)s]")
        parser.add_argument("-n", dest="dry_run", action="store_true",
                            default=False,
                            help="don't perform actions [default: %(default)s]"
                            )
        parser.add_argument("-r", "--framerate", dest="target_fps", default=30,
                            help="output framerate [default: %(default)s]")
        parser.add_argument("-R", "--renderer", dest="renderer", default=0, type=int, choices=range(len(RENDERERS)),
                            help="which renderer to use to display bars (0=filled, 1=hollow, "
                                 "2=symetrical filled, 3=symetrical hollow)")
        parser.add_argument("-v", "--version", action="version", version=program_version_string)

        parser.add_argument("-w", "--bar-width", dest="bar_width", default=5, type=int,
                            help="bar width in output images")
        parser.add_argument("-s", "--bar-spacing", dest="bar_spacing", default=2, type=int,
                            help="bar spacing in output images")
        parser.add_argument("-c", "--bar-count", dest="bar_count", default=100, type=int,
                            help="number of bars in output images")
        parser.add_argument("-C", "--color", dest="color", default='FFFFFFFF', type=str,
                            help="hexa color of bars [RRGGBB or RRGGBBAA, default: %(default)s]")
        parser.add_argument("-b", "--blending", dest="blending", default=0.7, type=float,
                            help="blending of previous spectrum into current one "
                                 "(0 = display only fresh data, 1 = use as many previous than fresh data)")
        parser.add_argument("-W", "--window", dest="fft_window", default=4096, type=int,
                            help="window size for FFT [default: %(default)s]")

        parser.add_argument("--image-height", dest="image_height", default=120, type=int, help="output images height")

        parser.add_argument("--audio-min-freq", dest="audio_min_freq", default=50, type=int,
                            help="min frequency in input audio")
        parser.add_argument("--audio-max-freq", dest="audio_max_freq", default=2500, type=int,
                            help="max frequency in input audio")
        parser.add_argument("--silence-ceiling", dest="silence_ceiling", default=70, type=float,
                            help="opposite of threshold considered silence [in dB, default: %(default)s]")

        parser.add_argument("-i", dest="input_filename", default="input.wav", help="input file in wav format")
        parser.add_argument("-o", dest="output_filename_mask", required=True,
                            help="output filename mask (should contain {:06} or similar to generate sequence)")

        # process options
        opts = parser.parse_args(argv)

    except Exception as e:
        indent = len(program_name) * " "
        sys.stderr.write(program_name + ": " + repr(e) + "\n")
        sys.stderr.write(indent + "  for help use --help")
        return 2

    if opts.debug:
        logging.root.setLevel(logging.DEBUG)
    else:
        logging.root.setLevel(logging.INFO)

    # convert hexa in input to RGBA tuple
    opts.color = (
        int(opts.color[0:2], 16),
        int(opts.color[2:4], 16),
        int(opts.color[4:6], 16),
        int(opts.color[6:8], 16) if len(opts.color) > 6 else 255
    )

    # load input file
    fs, data = wavfile.read(opts.input_filename)

    # compute bits per sample
    normalize_bits = 8 * data.dtype.itemsize - 1

    # monoize potential stereo files
    if len(data.shape) > 1:
        raw_data = data.T[0] + data.T[1]
        normalize_bits += 1
    else:
        raw_data = data.T

    # and normalize audio file in the [ -1 ; 1 ] range
    normalized_data = raw_data / 2 ** normalize_bits

    byte_per_frame = fs / opts.target_fps

    frame_start = 0
    frame_index = 0
    previous_spectrum = None

    frame_index_max = math.ceil(len(normalized_data) / byte_per_frame)

    writer = RENDERERS[opts.renderer](opts)

    while frame_start < len(normalized_data):
        # rms = np.sqrt(np.mean(np.square(normalized_data[frame_start: frame_start + 4096])))

        print("{0:6}/{1:6}".format(frame_index, frame_index_max), end="\r")

        # compute the raw spectrum
        spectrum = abs(np.fft.rfft(normalized_data[frame_start:frame_start + opts.fft_window]))

        # smooth it over time
        spectrum = smooth_spectrum(spectrum, previous_spectrum, opts.blending)

        # compute frequencies
        frequencies = compute_frequencies(spectrum, fs)

        # write spectrum to file
        writer.write_spectrum(frequencies, spectrum, frame_index)

        frame_start += byte_per_frame
        frame_index += 1
        previous_spectrum = spectrum

    return 0


if __name__ == "__main__":
    sys.exit(main())

# ffmpeg -i ~/starTrailsLoop_720p.mp4 -framerate 30 -i audio-00%4d.png -filter_complex "overlay=(main_w-overlay_w)/2:(main_h-overlay_h)/2:shortest=1" -i ~/Musique/change_wip.wav -map 2:0 -vframes 1151 -strict -2 test-b0.5.mp4 -y
#  ffmpeg -loop 1 -i ~/Téléchargement/star-tracks-1247850_1280.jpg -framerate 30 -i audio-00%4d.png -filter_complex "overlay=(main_w-overlay_w)/2:(main_h-overlay_h)/2:shortest=0" -i ~/Musique/change_wip.wav -map 2:0 -strict -2 -vframes 1151 test-b0.5.mp4 -y
